### ISSUE TITLE
GCE internal load balancer does not create ForwardingRule "api-[cluster_name]"

### ISSUE THREAD

**Issue by codingkriggs** [kubernetes/kops#15595]
/kind bug

This is my second ticket (just mentioning it here so you don't assume you're rereading the same ticket even though it starts the same way: https://github.com/kubernetes/kops/issues/15590).

Note, I have used substitutes `<bucket>` and `<project>` to hide private information in this report.

**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

`Client version: 1.26.4 (git-v1.26.4)`

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

```
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.3", GitCommit:"25b4e43193bcda6c7328a6d147b1fb73a33f1598", GitTreeState:"clean", BuildDate:"2023-06-14T09:53:42Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
The connection to the server localhost:8080 was refused - did you specify the right host or port?
```

**3. What cloud provider are you using?**

GCE

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

`$ kops create cluster simple.k8s.local --zones us-west1-c --state gs://<bucket>/ --project=<project> --api-loadbalancer-type internal --topology private --cloud gce --node-size e2-small --control-plane-size e2-small --yes`

**5. What happened after the commands executed?**

```
I0705 16:32:50.771288     971 new_cluster.go:1354] Cloud Provider ID: "gce"
I0705 16:32:50.954847     971 network.go:238] assigned 10.0.16.0/20 to subnet us-west1
W0705 16:32:56.762920     971 external_access.go:39] TODO: Harmonize gcemodel ExternalAccessModelBuilder with awsmodel
W0705 16:32:56.763091     971 firewall.go:41] TODO: Harmonize gcemodel with awsmodel for firewall - GCE model is way too open
W0705 16:32:56.769495     971 storageacl.go:165] adding bucket level write IAM for role "<bucket>" to gs://ControlPlane to support etcd backup
W0705 16:32:56.770983     971 autoscalinggroup.go:130] enabling storage-rw for etcd backups
W0705 16:32:56.771042     971 autoscalinggroup.go:296] Not hooking the instance group manager up to anything.
I0705 16:32:56.772994     971 executor.go:111] Tasks: 0 done / 67 total; 44 can run
W0705 16:32:56.816001     971 vfs_castore.go:383] CA private key was not found
I0705 16:32:56.830015     971 keypair.go:224] Issuing new certificate: "etcd-manager-ca-events"
I0705 16:32:56.830783     971 keypair.go:224] Issuing new certificate: "etcd-manager-ca-main"
I0705 16:32:56.845615     971 keypair.go:224] Issuing new certificate: "etcd-peers-ca-main"
I0705 16:32:56.864173     971 keypair.go:224] Issuing new certificate: "etcd-clients-ca"
I0705 16:32:56.884342     971 keypair.go:224] Issuing new certificate: "apiserver-aggregator-ca"
I0705 16:32:56.944786     971 keypair.go:224] Issuing new certificate: "etcd-peers-ca-events"
I0705 16:32:56.947589     971 healthcheck.go:111] GCE creating healthcheck: "api-simple-k8s-local"
W0705 16:32:56.965103     971 vfs_castore.go:383] CA private key was not found
I0705 16:32:57.025413     971 keypair.go:224] Issuing new certificate: "service-account"
I0705 16:32:57.045553     971 keypair.go:224] Issuing new certificate: "kubernetes-ca"
W0705 16:33:06.767337     971 executor.go:139] error running task "ProjectIAMBinding/serviceaccount-control-plane" (9m50s remaining to succeed): error updating IAM for project <project>: googleapi: Error 409: There were concurrent policy changes. Please retry the whole read-modify-write with exponential backoff., aborted
W0705 16:33:06.767383     971 executor.go:139] error running task "StorageBucketIAM/objectadmin-<bucket>-serviceaccount-controlplane" (9m50s remaining to succeed): error updating GCS bucket IAM for gs://<bucket>: googleapi: Error 400: Service account control-plane-simple-k8s-local@<project>.iam.gserviceaccount.com does not exist., invalid
I0705 16:33:06.767418     971 executor.go:111] Tasks: 42 done / 67 total; 17 can run
I0705 16:33:19.997288     971 executor.go:111] Tasks: 59 done / 67 total; 5 can run
I0705 16:33:33.136607     971 executor.go:111] Tasks: 64 done / 67 total; 2 can run
I0705 16:33:50.491958     971 executor.go:111] Tasks: 66 done / 67 total; 1 can run
I0705 16:33:50.604765     971 backend_service.go:127] GCE creating backend service: "api-simple-k8s-local"
I0705 16:34:03.301976     971 executor.go:111] Tasks: 67 done / 67 total; 0 can run
I0705 16:34:03.724811     971 update_cluster.go:323] Exporting kubeconfig for cluster
Error: error getting ingress status: error getting ForwardingRule "api-simple-k8s-local": googleapi: Error 404: The resource 'projects/<project>/regions/us-west1/forwardingRules/api-simple-k8s-local' was not found, notFound
```

**6. What did you expect to happen?**

The error on the last line of the log above should not have happened (saying forwarding rule api-simple-k8s-local was not found).
I assume that it should have been created at an earlier step in the process but never happened. I don't know how to confirm whether it exists after all is said and done, but this output makes me think it doesn't.

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```
$ kops get --name simple.k8s.local -o yaml
W0705 17:27:46.588701    1362 get.go:78] `kops get [CLUSTER]` is deprecated: use `kops get all [CLUSTER]`
```

```yaml
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: "2023-07-05T17:21:54Z"
  name: simple.k8s.local
spec:
  api:
    loadBalancer:
      type: Internal
  authorization:
    rbac: {}
  channel: stable
  cloudConfig: {}
  cloudProvider: gce
  configBase: gs://<bucket>/simple.k8s.local
  etcdClusters:
  - cpuRequest: 200m
    etcdMembers:
    - instanceGroup: control-plane-us-west1-c
      name: c
    memoryRequest: 100Mi
    name: main
  - cpuRequest: 100m
    etcdMembers:
    - instanceGroup: control-plane-us-west1-c
      name: c
    memoryRequest: 100Mi
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubeProxy:
    enabled: false
  kubelet:
    anonymousAuth: false
  kubernetesApiAccess:
  - 0.0.0.0/0
  - ::/0
  kubernetesVersion: 1.26.6
  networking:
    cilium:
      enableNodePort: true
  nonMasqueradeCIDR: 100.64.0.0/10
  project: <project>
  sshAccess:
  - 0.0.0.0/0
  - ::/0
  subnets:
  - cidr: 10.0.16.0/20
    name: us-west1
    region: us-west1
    type: Private
  topology:
    dns:
      type: Private
    masters: private
    nodes: private

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-07-05T17:21:54Z"
  labels:
    kops.k8s.io/cluster: simple.k8s.local
  name: control-plane-us-west1-c
spec:
  image: ubuntu-os-cloud/ubuntu-2004-focal-v20230302
  machineType: e2-small
  maxSize: 1
  minSize: 1
  role: Master
  subnets:
  - us-west1
  zones:
  - us-west1-c

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-07-05T17:21:54Z"
  labels:
    kops.k8s.io/cluster: simple.k8s.local
  name: nodes-us-west1-c
spec:
  image: ubuntu-os-cloud/ubuntu-2004-focal-v20230302
  machineType: e2-small
  maxSize: 1
  minSize: 1
  nodeLabels:
    cloud.google.com/metadata-proxy-ready: "true"
  role: Node
  subnets:
  - us-west1
  zones:
  - us-west1-c
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

(First I will delete the cluster, and then recreate it with the verbose flag.)

```
$kops delete cluster simple.k8s.local --yes
I0705 17:28:49.213458    1368 delete_cluster.go:128] Looking for cloud resources to delete
I0705 17:28:49.432278    1368 gce.go:99] Scanning zones: [us-west1-b us-west1-c us-west1-a]
TYPE                    NAME                                                                    ID
BackendService          api-simple-k8s-local                                                    api-simple-k8s-local
Disk                    c-etcd-events-simple-k8s-local                                          c-etcd-events-simple-k8s-local
Disk                    c-etcd-main-simple-k8s-local                                            c-etcd-main-simple-k8s-local
FirewallRule            master-to-master-simple-k8s-local                                       master-to-master-simple-k8s-local
FirewallRule            master-to-node-simple-k8s-local                                         master-to-node-simple-k8s-local
FirewallRule            node-to-master-simple-k8s-local                                         node-to-master-simple-k8s-local
FirewallRule            node-to-node-simple-k8s-local                                           node-to-node-simple-k8s-local
FirewallRule            nodeport-external-to-node-ipv6-simple-k8s-local                         nodeport-external-to-node-ipv6-simple-k8s-local
FirewallRule            nodeport-external-to-node-simple-k8s-local                              nodeport-external-to-node-simple-k8s-local
FirewallRule            ssh-external-to-master-ipv6-simple-k8s-local                            ssh-external-to-master-ipv6-simple-k8s-local
FirewallRule            ssh-external-to-master-simple-k8s-local                                 ssh-external-to-master-simple-k8s-local
FirewallRule            ssh-external-to-node-ipv6-simple-k8s-local                              ssh-external-to-node-ipv6-simple-k8s-local
FirewallRule            ssh-external-to-node-simple-k8s-local                                   ssh-external-to-node-simple-k8s-local
HealthCheck             api-simple-k8s-local                                                    api-simple-k8s-local
Instance                control-plane-us-west1-c-rkrp                                           us-west1-c/control-plane-us-west1-c-rkrp
Instance                nodes-us-west1-c-pj3w                                                   us-west1-c/nodes-us-west1-c-pj3w
InstanceGroupManager    c-control-plane-us-west1-c-simple-k8s-local                             us-west1-c/c-control-plane-us-west1-c-simple-k8s-local
InstanceGroupManager    c-nodes-us-west1-c-simple-k8s-local                                     us-west1-c/c-nodes-us-west1-c-simple-k8s-local
InstanceTemplate        control-plane-us-west1-c--s8llca-1688577745                             control-plane-us-west1-c--s8llca-1688577745
InstanceTemplate        nodes-us-west1-c-simple-k-pvjmo3-1688577745                             nodes-us-west1-c-simple-k-pvjmo3-1688577745
Network                 simple-k8s-local                                                        simple-k8s-local
Router                  nat-simple-k8s-local                                                    nat-simple-k8s-local
ServiceAccount          control-plane-simple-k8s-local@<project>.iam.gserviceaccount.com     projects/<project>/serviceAccounts/control-plane-simple-k8s-local@<project>.iam.gserviceaccount.com
ServiceAccount          node-simple-k8s-local@<project>.iam.gserviceaccount.com              projects/<project>/serviceAccounts/node-simple-k8s-local@<project>.iam.gserviceaccount.com
Subnet                  us-west1-simple-k8s-local                                               us-west1-simple-k8s-local

ServiceAccount:projects/<project>/serviceAccounts/node-simple-k8s-local@<project>.iam.gserviceaccount.comok
ServiceAccount:projects/<project>/serviceAccounts/control-plane-simple-k8s-local@<project>.iam.gserviceaccount.com       ok
HealthCheck:api-simple-k8s-local        error deleting resources, will retry: googleapi: Error 400: The health_check resource 'projects/<project>/regions/us-west1/healthChecks/api-simple-k8s-local' is already being used by 'projects/<project>/regions/us-west1/backendServices/api-simple-k8s-local', resourceInUseByAnotherResource
Disk:c-etcd-main-simple-k8s-local       ok
Disk:c-etcd-events-simple-k8s-local     ok
E0705 17:28:56.059454    1368 op.go:136] GCE operation failed: googleapi: Error 400: The instance_group_manager resource 'projects/<project>/zones/us-west1-c/instanceGroupManagers/c-control-plane-us-west1-c-simple-k8s-local' is already being used by 'projects/<project>/regions/us-west1/backendServices/api-simple-k8s-local'
InstanceGroupManager:us-west1-c/c-control-plane-us-west1-c-simple-k8s-local     error deleting resources, will retry: googleapi: Error 400: The instance_group_manager resource 'projects/<project>/zones/us-west1-c/instanceGroupManagers/c-control-plane-us-west1-c-simple-k8s-local' is already being used by 'projects/<project>/regions/us-west1/backendServices/api-simple-k8s-local'
FirewallRule:nodeport-external-to-node-ipv6-simple-k8s-local    ok
FirewallRule:nodeport-external-to-node-simple-k8s-local ok
Router:nat-simple-k8s-local     ok
FirewallRule:ssh-external-to-node-ipv6-simple-k8s-local ok
FirewallRule:ssh-external-to-master-simple-k8s-local    ok
FirewallRule:master-to-node-simple-k8s-local    ok
FirewallRule:ssh-external-to-node-simple-k8s-local      ok
FirewallRule:master-to-master-simple-k8s-local  ok
FirewallRule:ssh-external-to-master-ipv6-simple-k8s-local       ok
FirewallRule:node-to-node-simple-k8s-local      ok
FirewallRule:node-to-master-simple-k8s-local    ok
BackendService:api-simple-k8s-local     ok
Instance:us-west1-c/nodes-us-west1-c-pj3w       ok
Instance:us-west1-c/control-plane-us-west1-c-rkrp       ok
InstanceGroupManager:us-west1-c/c-nodes-us-west1-c-simple-k8s-local     ok
InstanceTemplate:nodes-us-west1-c-simple-k-pvjmo3-1688577745    ok
Not all resources deleted; waiting before reattempting deletion
        InstanceGroupManager:us-west1-c/c-control-plane-us-west1-c-simple-k8s-local
        Subnet:us-west1-simple-k8s-local
        HealthCheck:api-simple-k8s-local
        Network:simple-k8s-local
        InstanceTemplate:control-plane-us-west1-c--s8llca-1688577745
HealthCheck:api-simple-k8s-local        ok
InstanceGroupManager:us-west1-c/c-control-plane-us-west1-c-simple-k8s-local     ok
InstanceTemplate:control-plane-us-west1-c--s8llca-1688577745    ok
Subnet:us-west1-simple-k8s-local        ok
Network:simple-k8s-local        ok

Deleted cluster: "simple.k8s.local"
```

```
$ kops create cluster simple.k8s.local --zones us-west1-c --state gs://<bucket>/ --project=<project> --api-loadbalancer-type internal --topology private --cloud gce --node-size e2-small --control-plane-size e2-small --yes -v 10
```

See gist for output of this command: https://gist.github.com/codingkriggs/c351ba1bd18302c6a57019dc3ed38ab3

**9. Anything else do we need to know?**

I recently created another ticket that is a little bit related in that it involves private topology with GCE: https://github.com/kubernetes/kops/issues/15590

**Comment by hakman** [comment]
This should be fixed for 1.27.0. Thanks for reporting @codingkriggs!

**Comment by codingkriggs** [comment]
Confirmed: reported issue has been corrected in version 1.27.0

**Comment by hakman** [comment]
Thanks for checking @codingkriggs!
